### PR TITLE
sql: add new index_recommendation column

### PR DESCRIPTION
**Pull request by maryliag**
This commit adds a new column index_recommendations
STRING[] to:
crdb_internal.node_statement_statistics
crdb_internal.cluster_statement_statistics
system.statement_statistics
crdb_internal.statement_statistics

Part of #83782

Release note (sql change): Adding new column index_recommendations
to crdb_internal.node_statement_statistics,
crdb_internal.cluster_statement_statistics, system.statement_statistics
and crdb_internal.statement_statistics

### CODE REVIEW

**Review by postamar — DISMISSED** [review]
Indeed, at this time in the release cycle, avoiding these kinds of merge conflicts is a bit of a lost cause. Just a couple of minor comments on Schema's behalf.

> Reviewed 1 of 25 files at r1, 1 of 9 files at r2, 7 of 13 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84618#-:-N7pphXT2J9X58ufdVNa:b21yglo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @maryliag and @postamar)

___
*[`pkg/sql/catalog/systemschema/system.go` line 484 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84618#-N7pnDokCz9lw5UGIs5c:-N7pnDokCz9lw5UGIs5d:b-pucbst) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ca9d270b6abd842a56580253610b9a65ea3d3d8/pkg/sql/catalog/systemschema/system.go#L484)):*
> ```Go
>     ) STORED,
> 
>     index_recommendations STRING[],
> ```

Consider making this `NOT NULL DEFAULT (array[]::STRING[]))` instead? This way you avoid having to worry about NULL values altogether, which is always annoying with array types. You wouldn't need to add that function in `tree` either.
___
*[`pkg/upgrade/upgrades/alter_statement_statistics_index_recommendations.go` line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84618#-N7pomF91DNcP8BCt5IS:-N7pomFA38FzIjLb14Yw:b-g860id) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ca9d270b6abd842a56580253610b9a65ea3d3d8/pkg/upgrade/upgrades/alter_statement_statistics_index_recommendations.go#L31)):*
> ```Go
> func alterSystemStatementStatisticsAddIndexRecommendations(
> 	ctx context.Context, cs clusterversion.ClusterVersion, d upgrade.TenantDeps, _ *jobs.Job,
> ) error {
> ```

~It's worth adding a test for this upgrade step, just like the other upgrades. When spinning up your test cluster at version `clusterversion.AlterSystemStatementStatisticsAddIndexRecommendations`, do an `ALTER TABLE ... DROP COLUMN` to drop the `index_recommendations` column from the bootstrapped table to pretend that the table was created when bootstrapping from an earlier binary. Otherwise you're just testing the idempotency of the `addIndexRecommendationsCol`statement.~

edit: me not use reviewable gud, test is there and lgtm

**Review by mgartner — APPROVED** [review]
LGTM on Queries's (Query's?) behalf.

> Reviewed 12 of 25 files at r1, 4 of 9 files at r2, 13 of 13 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84618#-:-N7pzImJ0unVVejOhKC-:bmfb3w6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @maryliag)

**Review by postamar — COMMENTED** [review]
I found a few minor inconsistencies, not sure how they relate to the current test failures. Otherwise I have nothing to complain about on Schema's behalf.

> Reviewed 17 of 17 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84618#-:-N7rJHt0AVPpFs_Aclsu:b-gvnpp1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @maryliag)

___
*[`pkg/sql/catalog/systemschema/system.go` line 2015 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84618#-N7rIwte9kk3G5WrFXpn:-N7rIwte9kk3G5WrFXpo:b-27xnmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d93614a372d69add92fe78055645d5b52e015563/pkg/sql/catalog/systemschema/system.go#L2015)):*
> ```Go
> 					Hidden:      true,
> 				},
> 				{Name: "index_recommendations", ID: 12, Type: types.StringArray, Nullable: false},
> ```

Doesn't this also need some default expr field set now?
___
*[`pkg/upgrade/upgrades/alter_statement_statistics_index_recommendations.go` line 25 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84618#-N7rJ72dDA9CRWUO662u:-N7rJ72dDA9CRWUO662v:b-tzhxfq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d93614a372d69add92fe78055645d5b52e015563/pkg/upgrade/upgrades/alter_statement_statistics_index_recommendations.go#L25)):*
> ```Go
> const addIndexRecommendationsCol = `
> ALTER TABLE system.statement_statistics
> ADD COLUMN IF NOT EXISTS "index_recommendations" STRING[] DEFAULT (array[]::STRING[]))
> ```

Isn't NOT NULL missing here also?

**Review by maryliag — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84618)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @postamar)

___
*[`pkg/sql/catalog/systemschema/system.go` line 484 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84618#-N7pnDokCz9lw5UGIs5c:-N7uf-yx2eALAQfxSrj5:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ca9d270b6abd842a56580253610b9a65ea3d3d8/pkg/sql/catalog/systemschema/system.go#L484)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Consider making this `NOT NULL DEFAULT (array[]::STRING[]))` instead? This way you avoid having to worry about NULL values altogether, which is always annoying with array types. You wouldn't need to add that function in `tree` either.
</blockquote></details>

Done
___
*[`pkg/sql/catalog/systemschema/system.go` line 2015 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84618#-N7rIwte9kk3G5WrFXpn:-N7ufRJZ8BYd5Ziikegk:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d93614a372d69add92fe78055645d5b52e015563/pkg/sql/catalog/systemschema/system.go#L2015)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Doesn't this also need some default expr field set now?
</blockquote></details>

Done
___
*[`pkg/upgrade/upgrades/alter_statement_statistics_index_recommendations.go` line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84618#-N7pomF91DNcP8BCt5IS:-N7ufWy539AAtmWAeaOv:b2b0lyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ca9d270b6abd842a56580253610b9a65ea3d3d8/pkg/upgrade/upgrades/alter_statement_statistics_index_recommendations.go#L31)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

It's worth adding a test for this upgrade step, just like the other upgrades. When spinning up your test cluster at version `clusterversion.AlterSystemStatementStatisticsAddIndexRecommendations`, do an `ALTER TABLE ... DROP COLUMN` to drop the `index_recommendations` column from the bootstrapped table to pretend that the table was created when bootstrapping from an earlier binary. Otherwise you're just testing the idempotency of the `addIndexRecommendationsCol`statement.
</blockquote></details>

Talked offline, the test was already added
___
*[`pkg/upgrade/upgrades/alter_statement_statistics_index_recommendations.go` line 25 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84618#-N7rJ72dDA9CRWUO662u:-N7ufU785gEkkJFYCMrz:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d93614a372d69add92fe78055645d5b52e015563/pkg/upgrade/upgrades/alter_statement_statistics_index_recommendations.go#L25)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Isn't NOT NULL missing here also?
</blockquote></details>

Done

File: pkg/sql/sqlstats/persistedsqlstats/flush.go [hunk 1]
```diff
@@ -465,7 +474,7 @@ func (s *PersistedSQLStats) insertStatementStats(
 ) (rowsAffected int, err error) {
 	insertStmt := `
 INSERT INTO system.statement_statistics
-VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
+VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
 ON CONFLICT (crdb_internal_aggregated_ts_app_name_fingerprint_id_node_id_plan_hash_transaction_fingerprint_id_shard_8,
              aggregated_ts, fingerprint_id, transaction_fingerprint_id, app_name, plan_hash, node_id)
```
Comment (THardy98) [inline pkg/sql/sqlstats/persistedsqlstats/flush.go:479]: Should we also check for conflict on index recommendations here as well?
Is there a case where we want to insert into an existing stats row if our index recommendations have changed?

**Review by THardy98 — COMMENTED** [review]
Left a small comment, not familiar with the schema changes in `pkg/upgrade`, but I imagine other 👀  have taken a look.

**Review by maryliag — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84618)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @postamar, and @THardy98)

___
*[`pkg/sql/sqlstats/persistedsqlstats/flush.go` line 479 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/84618#-N7sIvNT6zcA78BACbU8-r5-479:-N7usHL63vuG637gcmvF:blyn5i2) ([raw file](https://github.com/cockroachdb/cockroach/blob/748c4c0b7094e0fc9124620c68a91ec296675921/pkg/sql/sqlstats/persistedsqlstats/flush.go#L479)):*
<details><summary><i>Previously, THardy98 (Thomas Hardy) wrote…</i></summary><blockquote>

Should we also check for conflict on index recommendations here as well?
Is there a case where we want to insert into an existing stats row if our index recommendations have changed?
</blockquote></details>

the conflict is for the things that shouldn't change from execution to execution, and the index recommendations (like plan, statistics) can change. So you try to insert, if the "key" values already exists, it will returns there was a conflict, so then it will use the update statement, that will update the recommendation column

**Comment by THardy98** [comment]
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84618)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @postamar, and @THardy98)
> 
> _[`pkg/sql/sqlstats/persistedsqlstats/flush.go` line 479 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/84618#-N7sIvNT6zcA78BACbU8-r5-479:-N7usHL63vuG637gcmvF:blyn5i2) ([raw file](https://github.com/cockroachdb/cockroach/blob/748c4c0b7094e0fc9124620c68a91ec296675921/pkg/sql/sqlstats/persistedsqlstats/flush.go#L479)):_
> 
> Previously, THardy98 (Thomas Hardy) wrote…
> the conflict is for the things that shouldn't change from execution to execution, and the index recommendations (like plan, statistics) can change. So you try to insert, if the "key" values already exists, it will returns there was a conflict, so then it will use the update statement, that will update the recommendation column

Ahh okay, makes sense 👍

**Review by THardy98 — APPROVED** [review]
LGTM 👍

**Comment by maryliag** [comment]
TFTR!!

bors r+

**Pull request merged**